### PR TITLE
Added complete Plan button and instant planHeaderWidget update

### PR DESCRIPTION
**Pull request by Evil00787**
now updating plan widget instantly with route
and added complete plan button

### CODE REVIEW

File: lib/pages/child/task_in_progress_page.dart [hunk 1]
```diff
@@ -53,44 +53,50 @@ class _ChildTaskInProgressPageState extends State<ChildTaskInProgressPage> with
...
-									heroTag: "fabTasks",
-									backgroundColor: AppColors.childTaskColor,
-								),
-							),
-							_getFab(false, state)
-						],
-					),
+				return  WillPopScope(
+				  onWillPop: () async {
+						Navigator.of(context).pop({"plan" : state is TaskInstanceLoaded ? state.planInstance : null});
```
Comment (stasgora) [inline lib/pages/child/task_in_progress_page.dart:58]: Do we need this whole thing? `PlanInstanceCubit` is reloadable so it should update on it's own

File: lib/pages/plan/plan_instance_details_page.dart [hunk 2]
```diff
@@ -53,10 +62,20 @@ class _PlanInstanceDetailsPageState extends State<PlanInstanceDetailsPage> {
 					content: AppSegments(segments: _buildPanelSegments(state))
 				),
 				loadingBuilder: (context, state) => _getView(
-					planInstance: widget.initialPlanInstance,
+					planInstance: updatedUiPlanInstance ?? widget.initialPlanInstance,
 					content: Expanded(child: Center(child: AppLoader()))
 				),
-			)
+			),
+			floatingActionButton: widget.initialPlanInstance.state.ended || _wasPlanCompleted || !widget.showActions || (updatedUiPlanInstance!= null && updatedUiPlanInstance.state.ended) ?
```
Comment (stasgora) [inline lib/pages/plan/plan_instance_details_page.dart:69]: This logic could be done in cubit and returned in state as one bool

**Review by stasgora — CHANGES_REQUESTED** [review]

File: lib/logic/common/plan_instance_cubit.dart [hunk 3]
```diff
@@ -52,6 +55,39 @@ class PlanInstanceCubit extends ReloadableCubit {
...
+		}
+		if(isInProgress(planInstance.duration)) {
+			planInstance.duration.last.to =  TimeDate.now();
+			updates.add(_dataRepository.updatePlanInstanceFields(planInstance.id, state: PlanInstanceState.completed, duration: planInstance.duration));
+		}
+		else updates.add(_dataRepository.updatePlanInstanceFields(planInstance.id, state: PlanInstanceState.completed));
+		List<UITaskInstance> uiInstances = await _taskInstancesService.mapToUIModels(updatedTaskInstances);
+
+		Future.wait(updates);
+		emit(ChildTasksLoadSuccess(uiInstances, uiPlanInstance));
```
[the remark this answers is not in the document]
Comment (Evil00787) [reply lib/logic/common/plan_instance_cubit.dart:89]: Also timer will change... And then it gets complicated imo

File: lib/logic/common/plan_instance_cubit.dart [hunk 4]
```diff
@@ -86,6 +85,7 @@ class PlanInstanceCubit extends ReloadableCubit {
 		List<UITaskInstance> uiInstances = await _taskInstancesService.mapToUIModels(updatedTaskInstances);
 
 		Future.wait(updates);
+		uiPlanInstance = await _dataAggregator.loadPlanInstance(planInstance: planInstance);
```
Comment (stasgora) [inline lib/logic/common/plan_instance_cubit.dart:88]: It's not going to fetch the fields again unless you only pass it the `planInstanceId` param - so right now the duration should be wrong

File: lib/logic/child/task_instance/task_instance_cubit.dart [hunk 5]
```diff
@@ -28,8 +28,9 @@ class TaskInstanceCubit extends Cubit<TaskInstanceState> {
 	final NotificationService _notificationService = GetIt.I<NotificationService>();
 	final UIDataAggregator _dataAggregator = GetIt.I<UIDataAggregator>();
 
-	TaskInstanceCubit(this._taskInstanceId, this._activeUser) : super(TaskInstanceStateInitial());
+	TaskInstanceCubit(this._taskInstanceId, this._activeUser, this.uiPlanInstance) : super(TaskInstanceStateInitial());
 
+	UIPlanInstance uiPlanInstance;
 	TaskInstance taskInstance;
 	PlanInstance planInstance;
 	Task task;
```
Comment (stasgora) [inline lib/logic/child/task_instance/task_instance_cubit.dart:36]: All those should be private

File: lib/logic/child/task_instance/task_instance_cubit.dart [hunk 6]
```diff
@@ -28,8 +28,9 @@ class TaskInstanceCubit extends Cubit<TaskInstanceState> {
 	final NotificationService _notificationService = GetIt.I<NotificationService>();
 	final UIDataAggregator _dataAggregator = GetIt.I<UIDataAggregator>();
 
-	TaskInstanceCubit(this._taskInstanceId, this._activeUser) : super(TaskInstanceStateInitial());
+	TaskInstanceCubit(this._taskInstanceId, this._activeUser, this.uiPlanInstance) : super(TaskInstanceStateInitial());
 
+	UIPlanInstance uiPlanInstance;
```
Comment (stasgora) [inline lib/logic/child/task_instance/task_instance_cubit.dart:33]: This is much better that keeping it in the UI layer but we can do even better. Cubit state is the place to keep all the data, UI should only ever call functions on cubits, not access any of the member fields. You can add a `UIPlanInstance` to `TaskInstanceStateInitial` and access it from there.

File: lib/logic/common/plan_instance_cubit.dart [hunk 7]
```diff
@@ -20,22 +20,22 @@ class PlanInstanceCubit extends ReloadableCubit {
 	final DataRepository _dataRepository = GetIt.I<DataRepository>();
 	final TaskInstanceService _taskInstancesService = GetIt.I<TaskInstanceService>();
 	final UIDataAggregator _dataAggregator = GetIt.I<UIDataAggregator>();
-	final ObjectId _planInstanceId;
+	UIPlanInstance uiPlanInstance;
 
 	PlanInstance planInstance;
 
-	PlanInstanceCubit(this._planInstanceId, ModalRoute modalRoute) : super(modalRoute);
+	PlanInstanceCubit(this.uiPlanInstance, ModalRoute modalRoute) : super(modalRoute);
```
Comment (stasgora) [inline lib/logic/common/plan_instance_cubit.dart:27]: In this case it's going to be a little more tricky as I have not yet written any support for custom initial states in reloadable cubits. I will look into this tomorrow

**Review by stasgora — CHANGES_REQUESTED** [review]

File: lib/logic/common/plan_instance_cubit.dart (hunk 7 above)
Comment (stasgora) [reply lib/logic/common/plan_instance_cubit.dart:27]: I will leave that case for when I will be properly implementing the page model passing functionality as I see no elegant way of doing it temporarily right now. For now you can leave this cubit as-is.

File: lib/logic/child/task_instance/task_instance_cubit.dart [hunk 8]
```diff
@@ -28,99 +28,98 @@ class TaskInstanceCubit extends Cubit<TaskInstanceState> {
 	final NotificationService _notificationService = GetIt.I<NotificationService>();
 	final UIDataAggregator _dataAggregator = GetIt.I<UIDataAggregator>();
 
-	TaskInstanceCubit(this._taskInstanceId, this._activeUser, this.uiPlanInstance) : super(TaskInstanceStateInitial());
+	TaskInstanceCubit(this._taskInstanceId, this._activeUser, _uiPlanInstance) : super(TaskInstanceStateInitial(_uiPlanInstance));
```
Comment (stasgora) [inline lib/logic/child/task_instance/task_instance_cubit.dart:31]: No type :wink: 
```suggestion
	TaskInstanceCubit(this._taskInstanceId, this._activeUser, UIPlanInstance _uiPlanInstance) : super(TaskInstanceStateInitial(_uiPlanInstance));
```

**Review by stasgora — APPROVED** [review]

**Pull request merged**